### ISSUE TITLE
Support for `in` keyword as a type guard

### ISSUE THREAD

**Issue by a1q123456** [microsoft/pylance-release#2074]
Currently, the Pylance can deduce the type or let the developer specify the type explicitly of variables to `Literal` types and perform any check based on them, such as:

```python

foo = 'A' # foo has type: Literal['A']

bar = 'B' # bar has type: Literal['B']

MyConstraints = Literal['A', 'B']

foobar: MyConstraints = 'C' # error

my_val: str = ''

if my_val == 'A' or my_val == 'B': # a type guard
    foobar = my_val
```

But sometimes, we would like to use the keyword `in` to perform an in-range check, which is more readable than multiple `or` in the if condition clause. In this case, the Pylance is unable to deduce the type,  i.e.:
```python

if my_val in ('A', 'B'): # ('A', 'B') has type: tuple[Literal['A'], Literal['B']], but the Pylance doesn't consider this as a type guard
    foobar = my_val # error, but should be OK

```

I'm wondering if we could support this scenario.

**Comment by erictraut** [comment]
Yes, this isn't currently a supported type guard pattern. For a full list of supported type guard patterns, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#type-guards).

Each new type guard pattern requires very specific code. Each one adds complexity and can slow down overall type analysis performance, so we're somewhat conservative about adding support for new patterns. We do so when we receive signal that the pattern is sufficiently commonplace. We use two sources for this signal:

1. Is the pattern supported by other Python type checkers (especially mypy, which has been around the longest)?
2. Has the request been upvoted (liked) by a number of pylance/pyright users?

As you note above, pylance already supports the `== <literal>` type guard pattern. To my knowledge, no other Python type checkers (including the latest version of mypy) support this.

Mypy also doesn't support the `in <tuple of literals>` pattern.

I think it would be technically feasible to support this pattern if we decided to make the investment, but it would need to be very constrained to the specific case where the type of the right-hand expression is a tuple (not a subclass of a tuple, since `__contains__` behavior can be overridden) and all members of the tuple type are literal types. It wouldn't work with lists or other iterables, with unions of tuples, subclasses of tuples, mixtures of literals and non-literals, etc.

In summary, we're open to adding this support if a sufficient number of pylance/pyright users indicate that they would like to see it.

In the meantime, you can use the `== <literal>` pattern as you noted above.

**Comment by kumaraditya303** [comment]
I think if tuple's `__contains__` in typeshed https://github.com/python/typeshed/blob/9c2be9500ae81d12858c6fd29a0ca14d8f284401/stdlib/builtins.pyi#L737 would be  annotated with TypeGuard then it would not require any change in pyright as it already supports TypeGuard.

**Comment by erictraut** [comment]
I added support for this type guard pattern a while ago, but I forgot that there was an enhancement request open in pylance. Closing now.
